File: Python/ByTier/D/1695_D1_Tree_Queries_Easy_Version.py
# Problem: CF 1695 D1 - Tree Queries (Easy Version)
# https://codeforces.com/contest/1695/problem/D1

"""
Algorithm: Tree Diameter and Centroid Finding
This code determines the minimum number of queries required to identify a hidden vertex in a tree.
It uses the concept of tree diameter and centroid properties to compute the minimum number of queries needed.
Time Complexity: O(n) where n is the number of vertices in the tree
Space Complexity: O(n) for storing the adjacency list representation of the tree
"""

import io
import os
import sys

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline


def make_graph(n, m):
    # Build adjacency list representation of the tree
    x, s = [0] * (2 * m), [0] * (n + 3)
    for i in range(0, 2 * m, 2):
        u, v = map(int, input().split())
        s[u + 2] += 1
        s[v + 2] += 1
        x[i], x[i + 1] = u, v
    for i in range(3, n + 3):
        s[i] += s[i - 1]
    G = [0] * (2 * m)
    for i in range(2 * m):
        j = x[i] + 1
        G[s[j]] = x[i ^ 1]
        s[j] += 1
    return G, s


def bfs(s):
    # Perform BFS from vertex s to compute distances
    q = [s]
    dist = [inf] * (n + 1)
    dist[s] = 0
    for k in range(n):
        i = q[k]
        di = dist[i]
        for v in range(s0[i], s0[i + 1]):
            j = G[v]
            if dist[j] == inf:
                q.append(j)
                dist[j] = di + 1
    return q, dist


t = int(input())
ans = []
inf = pow(10, 9) + 1
for _ in range(t):
    n = int(input())
    if n == 1:
        # If there's only one vertex, no queries are needed
        ans0 = 0
        ans.append(ans0)
        continue
    G, s0 = make_graph(n, n - 1)
    s = -1
    # Find a vertex with degree >= 3 (potential centroid)
    for i in range(1, n + 1):
        if s0[i + 1] - s0[i] >= 3:
            s = i
            break
    if s == -1:
        # If no vertex has degree >= 3, it's a path
        # We need just 1 query to distinguish between two endpoints
        ans0 = 1
        ans.append(ans0)
        continue
    # Find farthest vertices from the chosen vertex s
    p, dist = bfs(s)
    dp = [0] * (n + 1)
    # Dynamic programming to compute the minimum queries needed
    for i in p[::-1]:
        dpi, u = 0, 0
        for v in range(s0[i], s0[i + 1]):
            j = G[v]
            if dist[i] < dist[j]:
                dpi += max(dp[j], u)
                u = max(u, min(dp[j], 1) ^ 1)
        dp[i] = dpi
    # The answer is the value at the centroid vertex
    ans0 = dp[s]
    ans.append(ans0)
sys.stdout.write("\n".join(map(str, ans)))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1695/1695_D1_Tree_Queries_Easy_Version.cpp
// Problem: CF 1695 D1 - Tree Queries (Easy Version)
// https://codeforces.com/contest/1695/problem/D1

/*
 * Problem: D1. Tree Queries (Easy Version)
 * 
 * Purpose: 
 *   Given an unrooted tree with n vertices, determine the minimum number of queries 
 *   needed to uniquely identify a hidden vertex x. Each query asks for the distance 
 *   from a vertex v to x. The solution uses the concept of centroid decomposition and 
 *   tree traversal to compute the minimum number of queries required.
 * 
 * Algorithm:
 *   - For each node in the tree, perform a DFS to calculate how many "leaves" can be 
 *     covered by a single query from that node, based on the structure of the tree.
 *   - The result is the minimum among all such computations.
 * 
 * Time Complexity:
 *   O(n^2) in the worst case due to the nested loop of checking all nodes and performing DFS.
 *   In practice, for trees with reasonable branching, it's much faster.
 * 
 * Space Complexity:
 *   O(n) for storing the tree adjacency list and auxiliary arrays.
 */

#include <stdio.h>
#include <algorithm>
#include <vector>

using namespace std;
const int N = 1e5 + 5;
int t, n, x, y, f[N], ans;
vector<int> v[N];

// DFS to compute the number of required queries for a specific root node
void dfs(int x, int fa) {
  int c = 0; // Count of children that are leaves (no further children)
  for (int y : v[x])
    if (y != fa)
      dfs(y, x), f[x] += f[y], c += !f[y]; // Recursively process and count leaves
  f[x] += max(c - 1, 0); // Adjust based on number of children that are leaves
}

signed main() {
  scanf("%d", &t);
  while (t--) {
    scanf("%d", &n), ans = 1e9;
    for (int i = 1; i <= n; i++)
      v[i].clear(); // Clear previous edges
    
    // Read edges and build adjacency list
    for (int i = 1; i < n; i++)
      scanf("%d%d", &x, &y), v[x].push_back(y), v[y].push_back(x);
    
    // Special case: only one node
    if (n == 1) {
      puts("0");
      continue;
    }
    
    // Try every node as a starting point and calculate the min queries
    for (int i = 1; i <= n; i++)
      if (v[i].size() < 2) { // Only consider nodes with less than 2 neighbors
        fill(f + 1, f + 1 + n, 0); // Reset f array
        dfs(i, 0); // Compute the number of queries needed
        ans = min(ans, f[i] + 1); // Update minimum
      }
    
    printf("%d\n", ans);
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/